source/backend/api/Services/DocumentService.cs

```diff
@@ -123,39 +123,45 @@ public async Task<DocumentUpdateResponse> UpdateDocumentAsync(DocumentUpdateRequ
+            // Retrieve the existing metadata and check if it needs to be updated.
+            ExternalResult<QueryResult<DocumentMetadata>> existingMetadata = await documentStorageRepository.GetDocumentMetadataAsync(updateRequest.MayanDocumentId);
 
-                List<DocumentMetadataUpdateModel> updates = new List<DocumentMetadataUpdateModel>();
-                List<DocumentMetadataUpdateModel> creates = new List<DocumentMetadataUpdateModel>();
-                List<DocumentMetadataUpdateModel> deletes = new List<DocumentMetadataUpdateModel>();
+            List<DocumentMetadataUpdateModel> updates = new List<DocumentMetadataUpdateModel>();
+            List<DocumentMetadataUpdateModel> creates = new List<DocumentMetadataUpdateModel>();
+            List<DocumentMetadataUpdateModel> deletes = new List<DocumentMetadataUpdateModel>();
 
-                foreach (var updateMetadata in updateRequest.DocumentMetadata)
-                {
-                    var existing = existingMetadata.Payload.Results.FirstOrDefault(x => x.MetadataType.Id == updateMetadata.MetadataTypeId);
+            foreach (var updateMetadata in updateRequest.DocumentMetadata)
+            {
+                var existing = existingMetadata.Payload.Results.FirstOrDefault(x => x.MetadataType.Id == updateMetadata.MetadataTypeId);
 
-                    if (existing == null)
-                    {
-                        if (!string.IsNullOrEmpty(updateMetadata.Value))
-                        {
-                            creates.Add(updateMetadata);
-                        }
-                    }
-                    else if (existing.Value != updateMetadata.Value && !string.IsNullOrEmpty(updateMetadata.Value))
-                    {
-                        updateMetadata.Id = existing.Id;
-                        updates.Add(updateMetadata);
-                    }
-                    else if (string.IsNullOrEmpty(updateMetadata.Value))
+                if (existing == null)
+                {
+                    if (!string.IsNullOrEmpty(updateMetadata.Value))
                     {
-                        updateMetadata.Id = existing.Id;
-                        deletes.Add(updateMetadata);
+                        creates.Add(updateMetadata);
                     }
                 }
+                else if (existing.Value != updateMetadata.Value && !string.IsNullOrEmpty(updateMetadata.Value))
+                {
+                    updateMetadata.Id = existing.Id;
+                    updates.Add(updateMetadata);
+                }
+                else if (string.IsNullOrEmpty(updateMetadata.Value))
+                {
+                    updateMetadata.Id = existing.Id;
+                    deletes.Add(updateMetadata);
+                }
+            }
+
+            var metadataUpdateSucessful = false;
 
+            // If there are no changes, mark the external update as successfull
```

Comment:
github got a bit confused formatting these changes, but it seems like the main change is that the if statement at line 127 in the previous version was removed in favour of this if statement below. Can you help me understand the reasoning for that change? it seems like that would cause documentStorageRepository.GetDocumentMetadataAsync to get executed where it wouldn't have been previously.

Reply:
This was the cause of the bug. 
Thge line " foreach (var updateMetadata in updateRequest.DocumentMetadata)" was expecting that sometimes the updateRequest would not be empty or null before checking the metadata differences. This is a wrong assumption. The front end always sends a documentMetadata field array even if there are no changes.

Now it will perform the proper diff on the set and get the correct added/modified/deleted count. If the count of any of those  are not zero, then update them on Mayan and use that flag to mark the update to continue.

The GetDocumentMetadataAsync call is necessary regardless given our pattern is not to trust the front end to mark changes